src/usdb_syncer/usdb_id_file.py

```diff
@@ -0,0 +1,463 @@
+):
+    """Invalid file format with URL missing a specific query parameter"""
+
+    missing_parameter: str
+    bad_url: str
+
+    def __str__(self) -> str:
+        return (
+            f"{super().__str__()}: missing query parameter '{self.missing_parameter}'"
+        )
+
+
+@attrs.define
+class UsdbIdFileRepeatedQueryParameterMalformedUrlFormatError(
+    UsdbIdFileMalformedUrlFormatError
+):
+    """Invalid file format with URL specific query parameter occurring multiple times"""
+
+    repeated_parameter: str
+    bad_url: str
+
+    def __str__(self) -> str:
+        return (
+            f"{super().__str__()}: repeated query parameter '{self.repeated_parameter}'"
+        )
+
+
+@attrs.define
+class UsdbIdFileInvalidQueryParameterMalformedUrlFormatError(
+    UsdbIdFileMalformedUrlFormatError
+):
+    """Invalid file format with URL having an invalid query parameter"""
+
+    invalid_parameter: str
+    bad_url: str
+
+    def __str__(self) -> str:
+        return (
+            f"{super().__str__()}: invalid query parameter '{self.invalid_parameter}'"
+        )
+
+
+@attrs.define
+class UsdbIdFileUnparsableQueryParameterMalformedUrlFormatError(
+    UsdbIdFileMalformedUrlFormatError
+):
+    """Invalid file format with URL having a query parameter that cannot be parsed"""
+
+    unparsable_parameter: str
+    bad_url: str
+
+    def __str__(self) -> str:
+        return (
+            f"{super().__str__()}: could not parse query parameter "
+            f"'{self.unparsable_parameter}'"
+        )
+
+
+@attrs.define
+class UsdbIdFileEmptyFileError(UsdbIdFileError):
+    """Files do not contain any USDB ID but were selected for import"""
+
+    def __str__(self) -> str:
+        return "empty file"
+
+
+@attrs.define
+class UsdbIdFileInvalidJsonError(UsdbIdFileError):
+    """failed to interpret file content as JSON"""
+
+    def __str__(self) -> str:
+        return "invalid JSON format"
+
+
+@attrs.define
+class UsdbIdFileEmptyJsonArrayError(UsdbIdFileError):
+    """file content is an emty JSON array"""
+
+    def __str__(self) -> str:
+        return "empty JSON array"
+
+
+@attrs.define
+class UsdbIdFileNoJsonArrayError(UsdbIdFileError):
+    """file content is valid JSON, but not an array"""
+
+    def __str__(self) -> str:
+        return "file does not contain a JSON array"
+
+
+@attrs.define
+class UsdbIdFileInvalidUsdbIdError(UsdbIdFileError):
+    """expected USDB ID string cannot be converted correctly"""
+
+    def __str__(self) -> str:
+        return "invalid USDB ID in file"
+
+
+@attrs.define
+class UnexpectedUsdbIdFileInvalidUsdbIdError(UsdbIdFileInvalidUsdbIdError):
+    """some unknown error around parsing an USDB ID string"""
+
+    def __str__(self) -> str:
+        return "unexpected error when parsing USDB ID(s)"
+
+
+@attrs.define
+class UsdbIdFileNoUrlFoundError(UsdbIdFileError):
+    """parser could not find an URL in file content"""
+
+    def __str__(self) -> str:
+        return "no URL found"
+
```

Comment:
Indeed, a lot of exceptions. I also thought of maybe putting them to an `exceptions.py`, but this would require an `usdb_id_file` directory, which probably is a bit too much. Shall I try to change it to use Enums or shall we leave it that way?

Reply:
As you like. Either way is fine by me.

Reply:
I prefer to leave it like it is, changing might introduce another issue :)